File: internal/agentstate/activity.go
package agentstate

import (
	"time"

	"github.com/lucascaro/hive/internal/wire"
)

// Agent activity: what the session's agent is doing inside a turn —
// which tool it is running and where it is in its own plan. See
// docs/design-docs/agent-activity.md.
//
// This lives on Machine rather than beside it on registry.Entry so it
// inherits the lifecycle that already exists: New builds a fresh one,
// attachSessionHooks replaces it wholesale on create/restart/revive,
// and deleting the entry frees it. There is no second place to clear,
// which is the whole reason the state machine owns it.
//
// Like the rest of Machine this is deliberately NOT concurrency-safe.
// The registry mutex is the single guard.

const (
	// ActivityRingCap bounds the per-session tool history. It dies with
	// the daemon like the PTY — there is no disk format.
	ActivityRingCap = 200
	// activityOpenCap bounds tool calls awaiting their end. A reporter
	// that dies mid-tool never sends the end, so without a cap this map
	// grows for the life of the session.
	activityOpenCap = 32
)

// openCall is a tool_start waiting for its tool_end.
type openCall struct {
	tool      string
	target    string
	startedAt time.Time
	planIdx   int
}

// activity is the per-session ring plus the latest plan snapshot.
type activity struct {
	ring []wire.ToolEvent
	plan []wire.PlanItem
	open map[string]openCall

	// delta is the tool event the most recent Apply produced, for the
	// ACTIVITY broadcast. It is NOT "the last ring entry": a tool_start
	// that carries a call_id goes into open rather than the ring (it
	// has not finished), so reading the ring would silently broadcast
	// nothing at all when a tool begins — clients would only ever learn
	// that tools had ended.
	delta    wire.ToolEvent
	hasDelta bool
}

// currentPlanIdx is the plan item the agent says it is on, or -1.
func (a *activity) currentPlanIdx() int {
	for i := range a.plan {
		if a.plan[i].Status == wire.PlanStatusActive {
			return i
		}
	}
	return -1
}

// push appends to the ring, evicting the oldest past the cap.
func (a *activity) push(ev wire.ToolEvent) {
	a.ring = append(a.ring, ev)
	if len(a.ring) > ActivityRingCap {
		// Copy down rather than reslice: reslicing would keep the whole
		// backing array alive for the life of the session.
		n := copy(a.ring, a.ring[len(a.ring)-ActivityRingCap:])
		a.ring = a.ring[:n]
	}
}

// evictOldestOpen drops the longest-running unfinished call. A Go map
// has no order, so "oldest" is defined explicitly by startedAt; at 32
// entries the scan is cheaper than maintaining a second index.
func (a *activity) evictOldestOpen() {
	var oldestID string
	var oldestAt time.Time
	for id, c := range a.open {
		if oldestID == "" || c.startedAt.Before(oldestAt) {
			oldestID, oldestAt = id, c.startedAt
		}
	}
	if oldestID != "" {
		delete(a.open, oldestID)
	}
}

// toolStart records a tool beginning. now is the DAEMON's clock, not
// the reporter's: durations must not be able to straddle a clock
// adjustment on the reporting side.
func (m *Machine) toolStart(ev Event, now time.Time) {
	idx := m.act.currentPlanIdx()
	// The tally is incremented at START, so a tool that never reports
	// an end still counts against the step that launched it.
	if idx >= 0 {
		m.act.plan[idx].Tools++
	}
	started := wire.ToolEvent{
		Tool:      ev.Tool,
		Target:    ev.Target,
		CallID:    ev.CallID,
		StartedAt: now.UTC().Format(time.RFC3339Nano),
		PlanIdx:   idx,
	}
	// The delta reports the tool as running — OK stays nil, which is
	// exactly what "in flight" means on the wire.
	m.act.delta, m.act.hasDelta = started, true

	if ev.CallID == "" {
		// Nothing to pair with. Record it as an already-closed entry so
		// the timeline still shows that the tool ran.
		m.act.push(started)
		return
	}
	if m.act.open == nil {
		m.act.open = make(map[string]openCall, activityOpenCap)
	}
	if _, exists := m.act.open[ev.CallID]; !exists && len(m.act.open) >= activityOpenCap {
		m.act.evictOldestOpen()
	}
	m.act.open[ev.CallID] = openCall{
		tool:      ev.Tool,
		target:    ev.Target,
		startedAt: now,
		planIdx:   idx,
	}
}

// toolEnd closes a tool call and pushes it to the ring. An end with no
// matching start is recorded unpaired rather than dropped: "this ran
// and we missed the beginning" is more useful than silence.
func (m *Machine) toolEnd(ev Event, now time.Time) {
	out := wire.ToolEvent{
		Tool:    ev.Tool,
		Target:  ev.Target,
		CallID:  ev.CallID,
		EndedAt: now.UTC().Format(time.RFC3339Nano),
		OK:      ev.OK,
		PlanIdx: -1,
	}
	if ev.CallID != "" {
		if open, ok := m.act.open[ev.CallID]; ok {
			delete(m.act.open, ev.CallID)
			out.StartedAt = open.startedAt.UTC().Format(time.RFC3339Nano)
			out.PlanIdx = open.planIdx
			// Both ends are the daemon's own clock, so this subtraction
			// is meaningful in a way reporter timestamps are not.
			if d := now.Sub(open.startedAt); d > 0 {
				out.DurationMS = d.Milliseconds()
			}
			if out.Tool == "" {
				out.Tool = open.tool
			}
			if out.Target == "" {
				out.Target = open.target
			}
		}
	}
	if out.PlanIdx < 0 {
		out.PlanIdx = m.act.currentPlanIdx()
	}
	m.act.push(out)
	m.act.delta, m.act.hasDelta = out, true
}

// setPlan replaces the plan wholesale, carrying each item's tool tally
// forward by matching text.
//
// Wholesale replacement is what the reporters give us — a Claude
// TodoWrite call carries the entire list every time — but a naive
// replacement would reset every tally on each fire, since TodoWrite
// fires repeatedly as the plan evolves. Matching by text is
// first-match-wins: an agent can emit two steps with identical text,
// and truncation at MaxPlanTextLen can make two long steps identical,
// so each old item is consumed at most once.
func (m *Machine) setPlan(items []wire.PlanItem) {
	if len(items) > wire.MaxPlanItems {
		items = items[:wire.MaxPlanItems]
	}
	old := m.act.plan
	used := make([]bool, len(old))
	next := make([]wire.PlanItem, 0, len(items))
	for _, it := range items {
		it.Text = truncatePlanText(it.Text)
		if !wire.PlanStatuses[it.Status] {
			// Coerced, not dropped: losing a step entirely is worse
			// than mislabelling one.
			it.Status = wire.PlanStatusPending
		}
		it.Tools = 0
		for i := range old {
			if !used[i] && old[i].Text == it.Text {
				it.Tools = old[i].Tools
				used[i] = true
				break
			}
		}
		next = append(next, it)
	}
	m.act.plan = next
}

// truncatePlanText caps a plan step at the wire limit, on a rune
// boundary — the text is whatever the agent wrote.
func truncatePlanText(s string) string {
	if len(s) <= wire.MaxPlanTextLen {
		return s
	}
	cut := wire.MaxPlanTextLen
	for cut > 0 && !utf8RuneStart(s[cut]) {
		cut--
	}
	return s[:cut]
}

// utf8RuneStart reports whether b begins a UTF-8 rune.
func utf8RuneStart(b byte) bool { return b&0xC0 != 0x80 }

// Activity returns the stored ring and the current plan, as copies —
// the caller marshals them onto the wire while the registry lock is
// held elsewhere, and handing out the live slices would let a later
// push alias them.
func (m *Machine) Activity() ([]wire.ToolEvent, []wire.PlanItem) {
	events := make([]wire.ToolEvent, len(m.act.ring))
	copy(events, m.act.ring)
	plan := make([]wire.PlanItem, len(m.act.plan))
	copy(plan, m.act.plan)
	return events, plan
}

// LastToolDelta returns the tool event the most recent Apply produced,
// for the ACTIVITY broadcast. ok is false when no tool event has been
// applied yet.
//
// A started-but-unfinished tool is reported here with a nil OK and no
// duration; it reaches the ring only when it ends.
func (m *Machine) LastToolDelta() (wire.ToolEvent, bool) {
	return m.act.delta, m.act.hasDelta
}

// planSummary is the compact form carried on SessionInfo: how far
// along, and what is running right now. Clients render a row from
// these three values alone and never read the ring.
func (m *Machine) planSummary() (done, total int, current string) {
	for i := range m.act.plan {
		if m.act.plan[i].Status == wire.PlanStatusDone {
			done++
		}
	}
	total = len(m.act.plan)
	// The most recently started unfinished call. Parallel tools mean
	// there can be several; the newest is the one worth naming.
	var newest time.Time
	for _, c := range m.act.open {
		if current == "" || c.startedAt.After(newest) {
			current, newest = c.tool, c.startedAt
		}
	}
	return done, total, current
}

File: cmd/hivegui/frontend/src/lib/activity.ts
// Agent activity (spec 416): the client half of the daemon's per-session
// tool ring and plan. Pure — the store in store/activity.ts holds the
// result, the renderers in components/activity/ read it.
//
// Wire types mirror internal/wire/control.go (ToolEvent, PlanItem,
// ActivityMsg). Daemon-owned and snake_case only, like SessionInfo's
// state fields.

export interface ToolEvent {
  tool: string;
  target?: string;
  call_id?: string;
  agent_id?: string;
  agent_type?: string;
  // RFC3339Nano, daemon clock. started_at is absent for an end that never
  // paired with a start.
  started_at?: string;
  ended_at?: string;
  duration_ms?: number;
  // Absent while the call is running.
  ok?: boolean;
  // Index of the plan step active when the call started; -1 for none.
  plan_idx: number;
}

export interface PlanItem {
  id?: string;
  text: string;
  status: 'pending' | 'active' | 'done';
  // Calls this step ran. Kept on the item by the daemon, so it stays
  // right after the ring has evicted them.
  tools?: number;
}

export interface ActivityMsg {
  session_id: string;
  // The whole ring when full, otherwise the one changed event.
  events?: ToolEvent[] | null;
  // null or absent: unchanged. []: emptied. Always meaningful when full.
  plan?: PlanItem[] | null;
  full?: boolean;
  // Daemon-clock instant past which a WORKING session's tier is silent.
  // Monotonic per session, so it also orders frames.
  stale_at?: string;
}

export interface SessionActivity {
  events: ToolEvent[];
  plan: PlanItem[];
  staleAt: string;
}

// Mirrors agentstate.ActivityRingCap / activityOpenCap. The client holds
// the same bound of completed calls, plus the running ones the daemon
// keeps outside its ring.
export const ACTIVITY_RING_CAP = 200;
export const ACTIVITY_OPEN_CAP = 32;

export function emptyActivity(): SessionActivity {
  return { events: [], plan: [], staleAt: '' };
}

const isDone = (e: ToolEvent) => e.ok !== undefined || e.ended_at !== undefined;

// A call is identified by its call_id. The few without one (an end that
// never paired) are identified by their content, so a second snapshot
// does not duplicate them.
function keyOf(e: ToolEvent): string {
  return e.call_id
    ? `id:${e.call_id}`
    : `ev:${e.tool}|${e.target ?? ''}|${e.started_at ?? ''}|${e.ended_at ?? ''}`;
}

function instant(s: string | undefined): number {
  const n = s ? Date.parse(s) : Number.NaN;
  return Number.isNaN(n) ? 0 : n;
}

const whenOf = (e: ToolEvent) => instant(e.started_at ?? e.ended_at);

// applyActivity folds one ACTIVITY frame into a session's state.
//
// A GET_ACTIVITY answer and the deltas are written by different daemon
// goroutines, so they arrive in either order. Two rules make that safe:
// events merge as a union by call — a snapshot adds and completes calls
// but never removes one the client already holds (it may be an end that
// raced ahead of the snapshot) — and a frame whose stale_at is older than
// the one already stored cannot replace the plan or stale_at.
export function applyActivity(
  prev: SessionActivity,
  msg: ActivityMsg,
): SessionActivity {
  const older =
    !!msg.stale_at &&
    !!prev.staleAt &&
    instant(msg.stale_at) < instant(prev.staleAt);

  let events = prev.events;
  if (msg.events && msg.events.length > 0) {
    const byKey = new Map(prev.events.map((e) => [keyOf(e), e]));
    for (const ev of msg.events) {
      const k = keyOf(ev);
      const had = byKey.get(k);
      // A late start never regresses a finished call.
      if (had && isDone(had) && !isDone(ev)) continue;
      byKey.set(k, ev);
    }
    events = cap([...byKey.values()].sort((a, b) => whenOf(a) - whenOf(b)));
  }

  let plan = prev.plan;
  if (!older) {
    if (msg.full) plan = msg.plan ?? [];
    else if (Array.isArray(msg.plan)) plan = msg.plan;
  }
  const staleAt = msg.stale_at && !older ? msg.stale_at : prev.staleAt;

  if (events === prev.events && plan === prev.plan && staleAt === prev.staleAt)
    return prev;
  return { events, plan, staleAt };
}

// Drops the oldest completed calls past the ring cap and the oldest open
// ones past the open cap. Input is sorted oldest first.
function cap(events: ToolEvent[]): ToolEvent[] {
  let done = events.filter(isDone).length - ACTIVITY_RING_CAP;
  let open = events.length - events.filter(isDone).length - ACTIVITY_OPEN_CAP;
  if (done <= 0 && open <= 0) return events;
  return events.filter((e) => {
    if (isDone(e)) return done-- <= 0;
    return open-- <= 0;
  });
}

// isStale: the heuristic tier has nothing current to show, and a working
// session whose tier has been silent past stale_at is not reporting. A
// session at rest is never stale — no reporter heartbeats, so silence at
// rest is what a healthy tier looks like.
export function isStale(
  state: string | undefined,
  source: string | undefined,
  staleAt: string,
  now: number,
): boolean {
  if (!source || source === 'heuristic') return true;
  return state === 'working' && !!staleAt && now > instant(staleAt);
}

// How long the tier has been silent past its deadline, for the age label.
export function staleForMs(staleAt: string, now: number): number {
  return staleAt ? now - instant(staleAt) : 0;
}

export interface SubagentCalls {
  agentId: string;
  agentType: string;
  calls: ToolEvent[];
}

export interface CallGroup {
  main: ToolEvent[];
  subagents: SubagentCalls[];
}

export interface StepGroup extends CallGroup {
  item: PlanItem;
  index: number;
  tools: number;
}

// groupTimeline nests calls under the plan step they ran in. Subagent
// calls group by agent id within the step: nothing on the wire links a
// subagent to the parent call that spawned it.
export function groupTimeline(
  plan: PlanItem[],
  events: ToolEvent[],
): { steps: StepGroup[]; unplanned: CallGroup } {
  const groups: CallGroup[] = plan.map(() => ({ main: [], subagents: [] }));
  const unplanned: CallGroup = { main: [], subagents: [] };
  for (const ev of events) {
    const g =
      ev.plan_idx >= 0 && ev.plan_idx < plan.length
        ? groups[ev.plan_idx]
        : unplanned;
    if (!ev.agent_id) {
      g.main.push(ev);
      continue;
    }
    let sub = g.subagents.find((s) => s.agentId === ev.agent_id);
    if (!sub) {
      sub = { agentId: ev.agent_id, agentType: ev.agent_type ?? '', calls: [] };
      g.subagents.push(sub);
    }
    sub.calls.push(ev);
  }
  return {
    steps: plan.map((item, index) => ({
      item,
      index,
      tools: item.tools ?? 0,
      ...groups[index],
    })),
    unplanned,
  };
}

export function formatAge(ms: number): string {
  const s = Math.max(0, Math.floor(ms / 1000));
  if (s < 60) return `${s}s`;
  if (s < 3600) return `${Math.floor(s / 60)}m`;
  return `${Math.floor(s / 3600)}h`;
}

export function formatDuration(ms: number): string {
  if (ms < 1000) return `${Math.round(ms)}ms`;
  if (ms < 60_000) return `${(ms / 1000).toFixed(1)}s`;
  return `${Math.floor(ms / 60_000)}m ${Math.floor((ms % 60_000) / 1000)}s`;
}
